adr/keycloak-adr.md

```diff
@@ -0,0 +1,236 @@
+#### Architectural Decision Record (ADR): Integration of Keycloak with GitLab for Authentication
+
+#### Status
+Proposed
+
+#### Date
+2024-03-05
+
+
+#### Introduction
+This ADR supports the further integration of Keycloak with GitLab to streamline authentication mechanisms, with the goal of enhancing security and user experience. The drive behind this initiative is to refine login processes, bolster security protocols, and ensure compliance with Information Assurance (IA) standards.
+
+**Objective**: Utilize Keycloak as the predominant Identity and Access Management (IAM) solution alongside GitLab, applying OAuth2 for authentication. This plan aims to consolidate user management and authentication, making credential management more straightforward and improving our security landscape.
+
+**Rationale**: The integration of authentication between Keycloak and GitLab is fundamental for facilitating seamless user access, elevating security via centralized oversight, and adhering to compliance standards. This commitment is built on our adherence to open authentication and authorization standards, such as OAuth2 and OpenID Connect (OIDC).
+
+```mermaid
+graph LR
+  A["User"] -->|1. Requests access to GitLab| C["GitLab"]
+  C -->|2. Redirects to Keycloak| B["Keycloak"]
+  B -->|3. Authenticates user| A
+  B -->|4. Sends authentication token to GitLab| C
+  C -->|5. Grants user access| A
+
+  subgraph "Authentication Flow"
+    B["Keycloak (Identity Provider)"]
+    C["GitLab (Service Provider)"]
+    A["User (Requester)"]
+  end
+```
+
+#### Decision Process
+The decision to further integrate Keycloak with GitLab was informed by evaluating several key factors and prerequisites:
+
+```mermaid
+graph TD
+    A[Start] --> B{License Requirements}
+    B -->|Available: GitLab Premium| C[Proceed with OAuth2/OIDC]
+    B -->|Not Available: GitLab Premium| D[Explore Alternatives]
+    C --> E{Is SAML Integration Needed?}
+    E -->|Yes| F[Defer SAML Integration to Future Phases]
+    E -->|No| G[Proceed with OAuth2/OIDC]
+    G --> H{Interest in Group Integration?}
```

Comment:
This seems odd - we should be able to gather these reqs now from @CloudBeard / @bburky

Reply:
Compliance wise, it dosen't matter if its SAML, OAuth2, or OIDC so whatever works best for users/us.

There are controls for SSO/MFA and using RBAC

- AC-2 (16): Account Management | Role-Based Schemes - This control requires the use of a role-based access scheme that is connected to the organization's mission and business functions.

- AC-3: Access Enforcement - This control requires the enforcement of access control policies based on user roles, including the enforcement of least privilege.

- IA-2: Identification and Authentication (Organizational Users) - This control requires the use of MFA for network access to privileged and non-privileged accounts.

- IA-2 (1): Identification and Authentication (Organizational Users) | Network Access to Privileged Accounts - This control requires the use of MFA for network access to privileged accounts.

- IA-2 (2): Identification and Authentication (Organizational Users) | Network Access to Non-Privileged Accounts - This control requires the use of MFA for network access to non-privileged accounts.